### PR TITLE
Finish refactoring of breeze-dependent pre-commits to use breeze shell 

### PR DESCRIPTION
**Pull request by potiuk**
ome of our pre-commits were using code from Breeze in rather complex
way - by inserting PYTHONPATH and importing code from there. That was
complex and brittle and with recent changes of ShelParam https://github.com/apache/airflow/pull/35801 those
precommits required more and more dependencies to be added to their
pre-commit virtualenvs.

The reason that it was done this way was the assumption that someone
might want to run pre-commits locally without having breeze installed,
but this assumption and use case is rather unlikely, becasue breeze
becomes more and more useful and used so we can safely assume that
anyone who wants to do pre-commits will also have breeze installed and
on path. And anyway to run those pre-commits you need to have breeze
CI image pulled and built, so you should generally have breeze to run
them.

This PR finishes the series of PRs implementing the refactor and
completes the refactor.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Review by eladkal — APPROVED** [review]
>As a bonus MyPy output is now coloured (also when run as part of git commit).

cool!

**Review by jscheffl — COMMENTED** [review]
Uff, I was looking over this and needed to make a break after ~15min of reading. This PR is quite complex and changes a lot of things in one batch.
I would have preferred to split the refactoring and changes into separate commits to be able to better understand the features or even make rather (3?) smaller PRs. 
But unfortunately it is all a single commit.

Before shooting in the wild, can you help (the reviewer) a bit what you have tested and what specifically during review we should take a look for? Critical things to test to ensure it does not only "work on my machine"?

**Comment by potiuk** [comment]
Sure - yeah. good idea. I will split it. It grew quite a bit more than initial refactor, but I think I can split it to logical chunks.

It started from the idea that it will be a "small refactor" but when I did it, I found out that I need to add a lot more options to`breeze shell` command to make it possible for pre-commits to use it. But I think I can attempt to split it to more logical separate chunks.

Let me turn it into a draft and start separating out things.

**Comment by potiuk** [comment]
That is also a good idea as a "learning" exercise for those who will be reviewing it - i will try to turn it into more of a `breeze tutorial` explaining those changes in more detail

**Comment by potiuk** [comment]
Marking as Draft so that it's not merged accidentally before #35862 gets merged , but generally speaking - the last commit is quite ready and "complete" for review.

**Comment by potiuk** [comment]
OK. Merged https://github.com/apache/airflow/pull/35862 so made it ready for review.

**Comment by potiuk** [comment]
BTW. If this is still too big I am happy to move pre-commit by pre-commit separately .

**Comment by potiuk** [comment]
I started to separate things out https://github.com/apache/airflow/pull/35873

File: scripts/ci/pre_commit/common_precommit_utils.py
Comment (jscheffl) [inline scripts/ci/pre_commit/common_precommit_utils.py]: I am fearing that we do not force a check here if the CI image is matching the requirements when we run pre-commit. If the image is outdated and therefore does not catch a problem, no complaint is raised.

Finally we would trust the CI to find this problem but I would like to catch errors early and local. Could we 
1. Either fail each test that runs in the container if the container is not up-to-date ...or...
2. Add a pre-commit check that the container is current? (1 fail only instead of 30+)

**Review by jscheffl — CHANGES_REQUESTED** [review]
I see one problem during review and actually I see a "real" problem that the pipeline in your run did not turn red. Because in my local run it failed in `pre-commit run -a` in:

```
scripts/ci/pre_commit/pre_commit_update_er_diagram.py:31: error: Argument
"project_name" to "run_command_via_breeze_shell" has incompatible type "None";
expected "str"  [arg-type]
        project_name=None,
                     ^~~~
Found 1 error in 1 file (checked 205 source files)
Error 1 returned
```
...and doing the manual check I can confirm that the (not changed) ER update script sets a `None` as parameter whereas the utils function requires a `str` as type.
Why has this not been catched in your CI run? Actually the CI also executes with `breeze static-checks --all-files` - is there another (hidden) bug that this was not blocking?

File: scripts/ci/pre_commit/common_precommit_utils.py
Comment (potiuk) [reply scripts/ci/pre_commit/common_precommit_utils.py]: 1. This is deliberate choice and reflects the situation of many of our users

I want to avoid that. In many cases the image is out-dated (while still working in 9X% of cases) and the user does not want to pull/rebuild it because pulling/rebuilding requires a lot of networking and tme. There are two cases for it:

a) when you are travelling/in-the-plane/no connectivity : I want to achieve the goal that if you are travelling and you want to work "as usual" you should not have to disable pre-commits when they are enabled. 

b) when you have weak connectivity you want to rebuild your image with new dependencies relatitvely rarely - this is always quite a bit network intensive operation, even if it is heavily optimized by caching. Many of our contributors (for example @ephraimbuddy  - our release manager ) live in the place where networking is much worse to what we are used to.

2. This is already added in a "soft" form - every time you run breeze locally there is a check running to see if some of the code that "implies" rebuilding of the image (Dockerfile, setup.* files, generated dependencies, pyprojet.tomle) have been modified since the last time you built the image. And you are gently asked to rebuild and warned that if you don't - weird things might happen. And when you press `y` - the image will be rebuilt then. which gives the user freedom on deciding when to rebuild the image. When they have "weird error" - they will rebuild EVENTUALLY. And this is our goal - to let the user an easy way to rebuild the image, but do not force them.

So the whole system is designed for eventual consistency, but also to leave freedom to the user to decide if they prefer to do thing "fast" (or at all even if they have no networking) with lass consistency or eventually "fully consistently. It's been a deliberate decision and design documented in this ADR: https://github.com/apache/airflow/blob/main/dev/breeze/doc/adr/0012-asking-user-for-confirmation.md

Comment (potiuk) [reply scripts/ci/pre_commit/common_precommit_utils.py]: >  see one problem during review and actually I see a "real" problem that the pipeline in your run did not turn red. Because in my local run it failed in pre-commit run -a in:


Interesting. I am separating it out now as a first "subset" of this PR and I will look into it.

Comment (potiuk) [reply scripts/ci/pre_commit/common_precommit_utils.py]: Oh yeah... Actually I got it in now in my local (separated) PR:

Screenshot 2023-11-26 at 23 37 22

Unfortunattely `mypy` checks are not fully deterministic. Depending onn the set of files passed to mypy, mypy sometimes produces different results when some "small" things change. This is because of Python's lack of full static-typing and dynamic Python nature. MyPy tries to figure out which types are there by looking into other Python code and tries to look at the ways how things are used/initialized and **sometimes** it figures out things wrongly. We cannot do much about it - other than periodically updating to newer versions of `mypy` which will get - hopefully - better and better in this heuristics/guessing.

Comment (potiuk) [reply scripts/ci/pre_commit/common_precommit_utils.py]: (and yes - the problem is that I changed it in one of the recent versions. the new PR will have no such problems.

Comment (jscheffl) [reply scripts/ci/pre_commit/common_precommit_utils.py]: We also had sometimes problems in mypy in our local CI, if you only check selectively the files changed but others depend on a function signature, then this might not be catched. Therefore we moved away from differential checks and always make full mypy checks. But as Airflow is huge this would also be a drawback.

One question that just jumps into my head - mypy checks are serial (probably from the past) - but we are running multiple - in the new setup via breeze is it still required to run them serial or could we have them in parallel as well? Mypy is running on a single core else...

For the other in (1) and (2) you convinced me.
I just took a look and there is a "verbose" option in pre-commit (https://pre-commit.com/#pre-commit-configyaml---hooks) such that still if all hooks pass, you could print a warning that the image is outdated - but don't block - in the benefit to make the issue visible in pre-commit as well.
(Maybe an option for a follow-up PR though)

Comment (jscheffl) [reply scripts/ci/pre_commit/common_precommit_utils.py]: TLDR - If the one mypy problem (is a 1-liner) is fixed, not passing a `None`I am OK to approve and then propose to push all other to a different PR.

Comment (potiuk) [reply scripts/ci/pre_commit/common_precommit_utils.py]: > We also had sometimes problems in mypy in our local CI, if you only check selectively the files changed but others depend on a function signature, then this might not be catched. Therefore we moved away from differential checks and always make full mypy checks. But as Airflow is huge this would also be a drawback.

Yes. Happens.. This is the speed vs. accuracy thing as well. I think we sometimes have to accept some inconsistencies in this case and hope for better support for type checking in `rypy` (or whatever the name will be for someone to rewrite mypy in rust - or maybe add `mypy` checks to ruff. Thought the big "issue" is that it requires to be run in CI image because of all the dependencies. From the experience here it works in 9% percent of cases, the number of time when there are such discrepancies is maybe 1  % or less cases (that's a bit impression and anecdotal one - I have no hard data) - but this is the feeling.

> One question that just jumps into my head - mypy checks are serial (probably from the past) - but we are running multiple - in the new setup via breeze is it still required to run them serial or could we have them in parallel as well? Mypy is running on a single core else...

Yeah. I tried several times to run it in parallel, but unfortunately it never worked well. Ideally we should use built-in parallelism but https://github.com/python/mypy/issues/933 - this issue is open since 2015 without much progress. It's been added 2 weeks ago to https://github.com/python/mypy/issues/16472 meta issue tracker so there is some hope.

Splitting it via built in parallelism and runnig several mypy instances in parallel makes however the problem from previous paragraph WAY WAY worse - you not only get wrong results with "differential" check but you also get different results when you run `--all-files` check - and pretty randomly changing from execution to execution. The set of files you have  "per run" will change every time you add a new file, and then it also depends on how manu CPUs you have. So using serial in this case was deliberate choice of "stability" vs. "speed". This time the stability won - becasue there you could see different set of problems in 5 different runs one-after-the-other. It was pretty much unusable. 

But maybe we should try again :)  - feel free to try it and maybe you can make it works :)


> I just took a look and there is a "verbose" option in pre-commit (https://pre-commit.com/#pre-commit-configyaml---hooks) such that still if all hooks pass, you could print a warning that the image is outdated - but don't block - in the benefit to make the issue visible in pre-commit as well.
(Maybe an option for a follow-up PR though)

Yeah. Good idea. We should do it for all "serially" run pre-comits with Breeze image.  Especially that it is really useful if you have error (and then we do not even need --verbose flag - pre-commit will automatically print the output when pre-commit fails. There is a bit of a problem there (and this is why I added `--quiet` flag in the previous PR - when you have "parallel" run (default), you will see the message repeated as many times as you have CPUS - which is rather spammy, but for mypy - absolutely we could do it. I will add it once I separate out mypy pre-commits in the next PR.

Comment (potiuk) [reply scripts/ci/pre_commit/common_precommit_utils.py]: > TLDR - If the one mypy problem (is a 1-liner) is fixed, not passing a NoneI am OK to approve and then propose to push all other to a different PR.

BTW. Thanks to that I found another problem - but it's going to be fixed in the first PR :)

**Comment by potiuk** [comment]
This is what's left :) ... Should be small to complete the change :)

File: dev/breeze/src/airflow_breeze/utils/run_utils.py
Comment (potiuk) [inline dev/breeze/src/airflow_breeze/utils/run_utils.py:385]: Not needed any more after this last PR from the series .

**Review by jscheffl — APPROVED** [review]
Thanks! LGTM!

**Pull request merged**
